### PR TITLE
Use GrandPrix Metric model

### PR DESCRIPTION
**Pull request by camillobruni**
- Use more explicit Metric objects with parent and child-metrics
- Update statistics.mjs to use more ES6 features

### CODE REVIEW

File: resources/metric.mjs [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import * as Statistics from "./statistics.mjs";
+
+/** Number of milliseconds in a single minute. */
+export const MILLIS_PER_MIN = 60 * 1000;
+
+export class Metric {
+    /** @param {string} name */
```
Comment (rniwa) [inline resources/metric.mjs:7]: I don't think we should have these comments in the code.

File: resources/metric.mjs [hunk 2]
```diff
@@ -0,0 +1,90 @@
+import * as Statistics from "./statistics.mjs";
+
+/** Number of milliseconds in a single minute. */
+export const MILLIS_PER_MIN = 60 * 1000;
```
Comment (rniwa) [inline resources/metric.mjs:4]: Please rename the variable to MILLISECONDS_PER_MINUTE and remove the comment.

File: resources/metric.mjs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    add(value) {
+        if (typeof value !== "number")
+            throw new Error(`Adding invalid value=${value} to metric=${this.name}`);
+        this.values.push(value);
+    }
+
+    /**
+     * Calculate aggregate metrics for collected values.
+     */
+    compute() {
```
Comment (rniwa) [inline resources/metric.mjs:77]: Please call this function computeAggregatedMetrics or something and remove the comment.

File: resources/statistics.mjs [hunk 4]
```diff
@@ -75,20 +75,20 @@ export function confidenceIntervalDelta(confidenceLevel, numberOfSamples, sum, s
...
-    var quantile = cdfForProbability[degreesOfFreedom - 1]; // The first entry is for the one degree of freedom.
+    const quantile = cdfForProbability[degreesOfFreedom - 1]; // The first entry is for the one degree of freedom.
     return (quantile * sampleStandardDeviation(numberOfSamples, sum, squareSum)) / Math.sqrt(numberOfSamples);
 }
 
 export function confidenceInterval(values, probability) {
-    var sum = sum(values);
-    var mean = sum / values.length;
-    var delta = confidenceIntervalDelta(probability || 0.95, values.length, sum, squareSum(values));
+    const sum_value = sum(values);
```
Comment (rniwa) [inline resources/statistics.mjs:89]: Nit: camelCase.

File: resources/metric.mjs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        this.mean = 0.0;
+        this.geomean = 0.0;
+        this.delta = 0.0;
+        this.percentDelta = 0.0;
+
+        this.sum = 0.0;
+        this.min = 0.0;
+        this.max = 0.0;
+
+        /** @type {number[]} */
```
Comment (rniwa) [inline resources/metric.mjs:23]: I don't think we should have these comments.

Comment (camillobruni) [reply resources/metric.mjs:23]: - I will remove them (that's literally just copy-pasted from the grandprix repo)
- Might be something for the future though, the type annotations are quite nice when working with vscode (and we could have typescript verification)

File: resources/metric.mjs (hunk 3 above)
Comment (camillobruni) [reply resources/metric.mjs:77]: DONE.

File: resources/statistics.mjs (hunk 4 above)
Comment (camillobruni) [reply resources/statistics.mjs:89]: Ups, too much python :)

File: resources/metric.mjs [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import * as Statistics from "./statistics.mjs";
 
 /** Number of milliseconds in a single minute. */
```
Comment (rniwa) [inline resources/metric.mjs:3]: I don't think we need the comment now that the variable is appropriately named.

File: resources/benchmark-runner.mjs [hunk 7]
```diff
@@ -118,6 +119,11 @@ export class BenchmarkRunner {
         this._suites = suites;
         this._client = client;
         this._page = null;
+        this._metrics = {
+            __proto__: null,
+            Total: new Metric("Total"),
+            Score: new Metric("Score", "rpm"),
```
Comment (rniwa) [inline resources/benchmark-runner.mjs:125]: Why are we using "rpm" here? Didn't we decide to move away from it?

**Review by rniwa — APPROVED** [review]

File: resources/metric.mjs (hunk 6 above)
Comment (camillobruni) [reply resources/metric.mjs:3]: removed.

File: resources/benchmark-runner.mjs (hunk 7 above)
Comment (camillobruni) [reply resources/benchmark-runner.mjs:125]: Ups, copy-pasta. Changed unit to "score" as well.

**Comment by camillobruni** [comment]
- Use more explicit Metric objects with parent and child-metrics
- Update statistics.mjs to use more ES6 features

**Pull request merged**